java-checks/src/main/java/org/sonar/java/checks/unused/utils/AnnotationFieldReferenceFinder.java

```diff
@@ -0,0 +1,184 @@
+ * See the Sonar Source-Available License for more details.
+ *
+ * You should have received a copy of the Sonar Source-Available License
+ * along with this program; if not, see https://sonarsource.com/license/ssal/
+ */
+package org.sonar.java.checks.unused.utils;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonar.java.checks.unused.UnusedPrivateFieldCheck;
+import org.sonar.plugins.java.api.tree.AnnotationTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+/**
+ * Finds usages of fields inside annotations.
+ * I.e. currently it is used in {@link UnusedPrivateFieldCheck} to determine whether a field that is not used anywhere in the traditional
+ * sense may actually be used inside a testing annotation.
+ * <p>
+ * E.g.
+ * <pre>{@code
+ * // This field is used in the @FieldSource annotation below.
+ * private static final List<Integer> field = List.of(1, 2, 3);
+ *
+ * @ParameterizedTest
+ * @FieldSource("field")
+ * void test(int input) {
+ *   // ...
+ * }
+ * </pre>
+ */
+public class AnnotationFieldReferenceFinder extends BaseTreeVisitor {
+  private static final Logger LOG = LoggerFactory.getLogger(AnnotationFieldReferenceFinder.class);
+
+  private final StringLiteralFinder stringLiteralFinder = new StringLiteralFinder();
+
+  // Stores all field names and associated variable trees for which no usages of the symbol have been found so far.
+  // Fields are removed from this map, as soon as a string literal is found inside an annotation that matches the name of the field.
+  // In other words, all fields in this map that remain after visiting all annotations are unused.
+  private final Map<String, VariableTree> fieldNameToVariableTree;
+  private boolean hasEnteredClassTree = false;
+
+  // We are intentionally not using a fully qualified name of the annotation class here, as it might be missing in contexts with incomplete
+  // semantics.
+  private static final String FIELD_SOURCE_METHOD_ANNOTATION = "FieldSource";
+  private static final String FIELD_SOURCES_METHOD_ANNOTATION = "FieldSources";
+
+  private AnnotationFieldReferenceFinder(HashMap<String, VariableTree> fieldNameToVariableTree) {
+    this.fieldNameToVariableTree = fieldNameToVariableTree;
+  }
+
+  /**
+   * Constructs an instance of this visitor that looks for references of the given fields inside annotations.
+   */
+  public static AnnotationFieldReferenceFinder findReferencesTo(Collection<VariableTree> fields) {
+    var fieldNameToVariableTree = new HashMap<String, VariableTree>(fields.size());
```

Comment:
Can we imagine a test where we provide `findReferencesTo` with a collection of 2 fields with similar names but coming from 2 different classes?

Reply:
I see your point. Because a class with fields with repeated names is rejected by  `javac` but accepted by `ecj`, we should probably keep the defensive code. But I would still try to add a test if that is not too costly because, while improbable, it is not impossible that a developer invokes the static method with bad values.